guilib/measure/classes/abc.py

```diff
@@ -0,0 +1,603 @@
+        """Set new settings for this instance.
+
+        Check and set settings. This method is used in
+        the setter of the settings property.
+
+        Parameters
+        ----------
+        new_settings : dict or ConfigParser or str or Path or ViPErLEEDSettings
+            The new settings.
+
+        Returns
+        -------
+        settings_valid : bool
+            True if the new settings given were accepted.
+
+        Raises
+        ------
+        TypeError
+            If new_settings is neither a dict, ConfigParser, str,
+            or Path
+        TypeError
+            If an element of the mandatory settings is None or
+            is not a Sequence with length <= 3.
+
+        Emits
+        -----
+        QObjectSettingsErrors.MISSING_SETTINGS
+            If the settings have not been found
+            or if new_settings was None.
+        QObjectSettingsErrors.INVALID_SETTINGS
+            If any element of the new_settings does
+            not fit the mandatory settings.
+        """
+        try:                                                                    # TODO: make method that searches through invalid for old values and replaces deprecated ones, make it a method of the ViPErLEEDSettings class
+            new_settings = ViPErLEEDSettings.from_settings(new_settings)
+        except (ValueError, NoSettingsError):
+            emit_error(self, QObjectSettingsErrors.MISSING_SETTINGS)
+            return False
+        invalid = self.are_settings_invalid(new_settings)
+        for missing, *info in invalid:
+            emit_error(self, QObjectSettingsErrors.INVALID_SETTINGS,
+                       missing, ' '.join(info))
+            return False
+
+        self._settings = new_settings
+        return True
+
+
+# DISABLE: seems a bug in pylint. This is also an ABC.
+# pylint: disable-next=abstract-method
+class HardwareABC(QObjectWithSettingsABC):
+    """Abstract base class of hardware-related objects."""
+
+    # Emitted whenever the busy state of the device changes.
+    # Contains the new busy state of the device.
+    busy_changed = qtc.pyqtSignal(bool)
+
+    def __init__(self, *args, **kwargs):
+        """Initialise instance."""
+        # Busy state of the device.
+        self._busy = False
+        super().__init__(*args, **kwargs)
+
+    @property
+    def busy(self):
+        """Return busy state of instance."""
+        return self.get_busy()
+
+    @busy.setter
+    def busy(self, is_busy):
+        """Set busy state of instance.
+
+        Note that 'self.busy =' will be performed in the calling
+        thread. To avoid this, one can invoke the set_busy slot.
+
+        Parameters
+        ----------
+        is_busy : bool
+            True if the instance is busy.
+
+        Emits
+        -----
+        busy_changed
+            If the busy state of the instance changed.
+        """
+        self.set_busy(is_busy)
+
+    def get_busy(self):
```

Comment:
Do we want this to be public to the outside (i.e., not only in subclasses)? Perhaps better to make it a `_get_busy` instead? The only reason to make it public would be to make it also a `pyqtSlot`.